src/Core/Content/Product/SalesChannel/Detail/ProductDetailRoute.php

```diff
@@ -184,12 +191,27 @@ private function findBestVariant(string $productId, SalesChannelContext $context
             ->addSorting(new FieldSorting('product.price'))
             ->setLimit(1);
 
-        $criteria->setTitle('product-detail-route::find-best-variant');
+        $criteria->setTitle('product-detail-route::find-best-variant-by-term');
         $variantId = $this->productRepository->searchIds($criteria, $context);
 
         return $variantId->firstId() ?? $productId;
     }
 
+    private function findBestVariantByTerm(string $term, string $productId, SalesChannelContext $context): ?string
```

Comment:
please override method `findBestVariant` with term

```suggestion
    private function findBestVariant(string $productId, SalesChannelContext $context, ?string $term): ?string
    {
    $criteria = (new Criteria())
            ->addFilter(new EqualsFilter('product.parentId', $productId))
            ->addSorting(new FieldSorting('product.available', FieldSorting::DESCENDING))
            ->addSorting(new FieldSorting('product.price'))
            ->setLimit(1);

        if ($term) {
           $criteria->addState(Criteria::STATE_ELASTICSEARCH_AWARE);
           $criteria->setTerm($term);
           $criteria->setTitle('product-detail-route::find-best-variant-by-term');
        }
        
        $variantId = $this->productRepository->searchIds($criteria, $context);

        return $variantId->firstId() ?? $productId;
    }
```

Reply:
that's also fine but you need to take a note on the sorting, that's the important part